### PR TITLE
[kuberay][autoscaler] Use new autoscaling fields from the KubeRay operator

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

This PR incorporates changes from KubeRay PRs https://github.com/ray-project/kuberay/pull/163 https://github.com/ray-project/kuberay/pull/274  https://github.com/ray-project/kuberay/pull/278 

- adds logic to process the new fields `upscalingMode` and `idleTimeoutSeconds` in the KubeRay CR
- refactors the example config to use kuberay's ability to autogenerate an autoscaler container
- uses the new `autoscalerOptions.image` field to set the autoscaler image in the test 
- uses a small `idleTimeoutSeconds` to test scale-down in the autoscaler e2e test

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Integration tests

### CODE REVIEW

File: python/ray/autoscaler/_private/kuberay/autoscaling_config.py
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/kuberay/autoscaling_config.py:89]: @harryge00 here's where we set upscaling speed

**Comment by DmitriGekhtman** [comment]
Hmm...getting some logs claiming the operator doesn't have permissions to list service accounts, but the RBAC looks as expected...need to debug more.

**Comment by DmitriGekhtman** [comment]
Ok, having the issue with the operator not being able to list service accounts in upstream KubeRay as well.
Also reproduced with some dummy objects in my K8s cluster. Maybe my K8s cluster is messed up somehow?

**Comment by DmitriGekhtman** [comment]
Ok...no, definitely not my K8s cluster...going to keep debugging.

**Comment by DmitriGekhtman** [comment]
Reviewers, please review -- I'm trying to push things along for the Ray 2.0.0beta release.

**Review by wuisawesome — CHANGES_REQUESTED** [review]
mostly minor comments about seeing if we can clean up the testing logic a little

File: python/ray/tests/kuberay/scripts/scale_down.py
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/scripts/scale_down.py:22]: Hmm, yeah, good point. I think it should be possible to express this as a context manager. Can add that to the test utils.

File: python/ray/tests/kuberay/scripts/scale_down.py
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/scripts/scale_down.py:32]: The issue is that wait for condition would throw without displaying an error message.

The solution is to refactor wait for condition a bit (rather than attempting to work around its current structure).

File: python/ray/autoscaler/kuberay/kuberay-autoscaler.patch
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/kuberay-autoscaler.patch:1]: It's generated manually with a git command.

Let me actually get rid of the patch -- we can do better by supplying a kustomize overlay in the Ray repo.

File: python/ray/tests/kuberay/scripts/scale_down.py
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/scripts/scale_down.py]: Added a decorator to test_utils, used it in the scale down and scale up scripts.

File: python/ray/tests/kuberay/scripts/scale_down.py
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/scripts/scale_down.py]: Took care of logging to stdout in a finally clause in the new decorator.
The condition is now a True/False lambda.

File: python/ray/tests/kuberay/scripts/scale_down.py
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/scripts/scale_down.py:14]: Redundancy fixed.

**Comment by DmitriGekhtman** [comment]
There's some incorrect downscaling behavior which is causing the test to fail. Going to debug.

**Comment by DmitriGekhtman** [comment]
Welcome, codeowner SWAT team.

File: python/ray/autoscaler/kuberay/kuberay-autoscaler.patch
Comment (wuisawesome) [reply python/ray/autoscaler/kuberay/kuberay-autoscaler.patch:1]: that seems better, but fine to do separate, i just wanted to make sure you weren't doing it by hand.

**Review by wuisawesome — APPROVED** [review]

File: python/ray/autoscaler/kuberay/kuberay-autoscaler.patch
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/kuberay-autoscaler.patch:1]: Will fix that in a future PR.

**Comment by DmitriGekhtman** [comment]
> There's some incorrect downscaling behavior which is causing the test to fail. Going to debug.

For once, autoscaler was correct, I (test) was wrong. Test fixed. Let's wait for CI to run.

**Comment by DmitriGekhtman** [comment]
Head pod failed to start in the CI, very strange...will probably have to debug the CI environment.

**Comment by DmitriGekhtman** [comment]
It's an issue with ImagePull policy for the autoscaler container. This will require another quick fix to the KubeRay repo.

**Comment by DmitriGekhtman** [comment]
...crossing fingers that it works and I don't have to debug it in the CI environment again....

**Comment by DmitriGekhtman** [comment]
Ah. There's the green check mark.
Will let the CI run to completion before merging.

**Comment by DmitriGekhtman** [comment]
100% green, except for the forsaken MacOS tests, which never run.
Merging.

**Pull request merged**
